04 - Guides, Workflows, & Courses/Guides/How to add automated tests to your plugin.md

```diff
@@ -1,18 +1,405 @@
+
+3. Pass in Obsidian API depending functionality as function parameter. Please note that `any` is not a good choice for a type - we'll take a look at this problem in [[How to add automated tests to your plugin#Using Obsidian API for typing in extracted business logic|Using Obsidian API for typing in extracted business logic]].
+
+```ts
+// myplugin.ts
+export default class MyPluginLogic {
+  static ribbonIconCallback(notice: any) {
+    // Called when the user clicks the icon.
+    new notice('This is a notice!');
+  }
+}
+```
+
+4. Adjust the callback in `main.ts` to use our extracted function
+
+```ts
+// main.ts
+import MyPluginLogic from "./myplugin";
+
+// [other code ...]
+const ribbonIconEl = this.addRibbonIcon('dice', 'Sample Plugin', (evt: MouseEvent) => {
+	MyPluginLogic.ribbonIconCallback(Notice)
+});
+```
+
+Restart Obsidian to check manually that the ribbon icon is still showing you the notice when clicked.
+
+> [!attention] Correct import path
+> It might happen that Visual Studio Code auto-imports `MyPluginLogic` like this:
+> ```ts
+> import MyPluginLogic from "myplugin";
+> ```
+> Mind the missing `./` in front of the file name. If your code is not working, try adjusting the import to `import MyPluginLogic from "./myplugin";`  
+
+Having the business logic extracted, it's now possible to put `myplugin.ts` under test instead of `main.ts` and avoid mocking obsidian API. 
+
+5. Remove or comment out `__mocks__/obsidian.ts` to make sure the following works without any mocks. 
+6. Replace the content of `example.spec.ts` with:
+
+```ts
+// example.spec.ts
+import MyPluginLogic from "./myplugin";
+
+describe("MyPlugin Tests", () => {
+  it('should create a notice with This is a notice! as text', () => {
+    const mockNotice = jest.fn();
+    MyPluginLogic.ribbonIconCallback(mockNotice)
+
+    expect(mockNotice).toHaveBeenCalledWith("This is a notice!")
+  })
+});
+```
+
+6. Run the test via `npm run test`. It should be green.
+
+Mind that you will not be able to test if you really registered a corresponding ribbonIcon with this approach. Generally, separating business logic from the logic thats necessary to integrate with Obsidian can help with readability and maintainability and might be valuable not only for testing. 
+
+##### Using Obsidian API for typing in extracted business logic
+
+As previously mentioned, using types of the Obsidian API is not a problem per se. To leverage the power of typescript and use the correct types, we can enhance `myplugin.ts` to correctly type the dependency the callback expects, even if this means having an Obsidian API import inside the code. 
+
+Since we're using a constructor here, you cannot just write `notice: Notice`, which would expect a already constructed Notice object; instead, use an [abstract construct signature](https://www.typescriptlang.org/docs/handbook/2/classes.html#abstract-construct-signatures).
+
+```ts
+// myplugin.ts
+import { Notice } from "obsidian";
```

Comment:
Maybe use explicit `import type { Notice } from "obsidian";` notation. Adding `"verbatimModuleSyntax": true,` to the TS config is a good idea anyways.

Reply:
You have a point, but I still would like to keep it this way. If one uses a obsidian dependency, i.e. `Notice`, VSC creates an auto-import like this. While a type import would be cleaner, it is not necessary to use them to get the tests running and I'd like to only use and mention mandatory steps to keep the complexity lower.

Reply:
That is fair. Though without explicit type imports, you are at the mercy of whatever the TS transpiler in this setup does.